SCClassLibrary/Common/Audio/BasicOpsUGen.sc

```diff
@@ -103,394 +130,653 @@ BinaryOpUGen : BasicOpUGen {
+					var sum4 = Sum4.newDuringOptimisation(this.rate, inputs[0], inputs[1], inputs[2], otherValues[0]);
+					add.prepareForReplacement(sum4, result, 1) !? { |re|
+						add.replace(re);
+					}
+				}
+			};
+			result.returnNilIfEmpty !? { |r| ^r };
+		};
+
+		// Sum3(a, b, c) - a => (b + c)
+		desc = descendants.select { |d| d.isKindOf(BinaryOpUGen) and: { d.operator == '-' } };
+		if(desc.isEmpty.not){
+			var result = SynthDefOptimizationResult();
+			desc.do { |minus|
+				var otherValues = minus.inputs.select{|i| i !== this };
+				// if Sum3(a, b, c) - Sum3(a, b, c) then the optimisation cannot be applied (handled by BinaryOpUGen)
+				if ((otherValues.size == 1) and: { inputs.includes(otherValues[0]) }) {
+					var inputCopy = inputs.copy;
+					var add;
+					inputCopy.remove(otherValues[0]);
+					add = BinaryOpUGen.newDuringOptimisation(this.rate, '+', inputCopy[0], inputCopy[1]);
+					minus.prepareForReplacement(add, result, 1) !? { |re|
+						minus.replace(re)
+					}
+				}
+			};
+			result.returnNilIfEmpty !? { |r| ^r };
+		};
+
+		if (inputs.every(_ === inputs[0])){
+			var result = SynthDefOptimizationResult();
+			var newMul = BinaryOpUGen.newDuringOptimisation(this.rate, '*', inputs[0], 3);
+			this.prepareForReplacement(newMul, result, 1) !? { |re|
+				this.replace(re);
+				^result
+			}
+		};
+
+		if (inputs.any(_ == 0)){
+			var nonZero = inputs.select(_ != 0);
+			case
+			{ nonZero.size == 0 } {
+				var result = SynthDefOptimizationResult();
+				this.prepareForReplacement(0, result, 1) !? { |re|
+					this.replace(re);
+					^result
+				}
+			}
+			{ nonZero.size == 1 } {
+				var result = SynthDefOptimizationResult();
+				this.prepareForReplacement(nonZero[0], result, 1) !? { |re|
+					this.replace(re);
+					^result
+				}
+			}
+			{ nonZero.size == 2 } {
+				var result = SynthDefOptimizationResult();
+				var new = BinaryOpUGen.newDuringOptimisation(this.rate, '+', nonZero[0], nonZero[1]);
+				this.prepareForReplacement(new, result, 1) !? { |re|
+					this.replace(re);
+					^result
+				}
+			}
+		};
+		^nil;
+	}
 }
 
 Sum4 : UGen {
+	implicitResourceConnectionStrategies { ^[] }
+	hasObservableEffect { ^false }
+	canBeReplacedByIdenticalCall { ^true }
+
 	*new { arg in0, in1, in2, in3;
 		^this.multiNew(nil, in0, in1, in2, in3)
 	}
 
 	*new1 { arg dummyRate, in0, in1, in2, in3;
-		var argArray, rate, sortedArgs;
+		var argArray = [in0, in1, in2, in3];
+		var rate = argArray.rate;
+		var sortedArgs = argArray.sort {|a b| a.rate < b.rate};
 
-		if (in0 == 0.0) { ^Sum3.new1(nil, in1, in2, in3) };
-		if (in1 == 0.0) { ^Sum3.new1(nil, in0, in2, in3) };
-		if (in2 == 0.0) { ^Sum3.new1(nil, in0, in1, in3) };
-		if (in3 == 0.0) { ^Sum3.new1(nil, in0, in1, in2) };
+		^super.new1(rate, *sortedArgs)
+	}
 
-		argArray = [in0, in1, in2, in3];
-		rate = argArray.rate;
-		sortedArgs = argArray.sort {|a b| a.rate < b.rate};
+	optimize {
+		var desc;
+
+		// If scalar or DC, actually do the maths.
```

Comment:
This might be worth a few methods?

```supercollider


+ Number {
	isDC { ^false }
	isConstantUGenInput { ^true }
}

+ UGen {
	isDC { ^this.source.isDC }
	isConstantUGenInput { ^false }
}

+ DC {
	isDC { ^true }
	isConstantUGenInput { ^true }
}

if (inputs.every(_.isConstant)) {
	var result = SynthDefOptimizationResult();
	var numbers = inputs.collect{ |in|
		if (in.isDC) { in.source.inputs[0] } { in }
	};
	var answer = numbers[0] + numbers[1] + numbers[2] + numbers[3];
	this.prepareForReplacement(answer, result, 1) !? { |replacement|
		this.replace(replacement);
		^result
	}
};
```